driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,234 @@
+use super::stablex_driver::StableXDriver;
+use crate::metrics::StableXMetrics;
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::{anyhow, Context, Result};
+use log::error;
+use log::info;
+use std::collections::HashSet;
+use std::thread;
+use std::time::{Duration, Instant, SystemTime, SystemTimeError};
+
+const BATCH_DURATION: Duration = Duration::from_secs(300);
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct BatchId(pub u64);
+
+impl BatchId {
+    pub fn current(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        let time_since_epoch = now.duration_since(SystemTime::UNIX_EPOCH)?;
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    pub fn start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    orderbook_reader: &'a dyn StableXOrderBookReading,
+    metrics: &'a StableXMetrics,
+    batch_wait_time: Duration,
+    max_batch_elapsed_time: Duration,
+    past_auctions: HashSet<BatchId>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum Action {
+    AlreadyHandled,
+    NotEnoughTimeLeft,
+    Solve,
+}
+
+impl<'a> Scheduler<'a> {
+    pub fn new(
+        driver: &'a mut dyn StableXDriver,
+        orderbook_reader: &'a dyn StableXOrderBookReading,
+        metrics: &'a StableXMetrics,
+        batch_wait_time: Duration,
+        max_batch_elapsed_time: Duration,
+    ) -> Self {
+        Self {
+            driver,
+            orderbook_reader,
+            metrics,
+            batch_wait_time,
+            max_batch_elapsed_time,
+            past_auctions: HashSet::new(),
+        }
+    }
+
+    pub fn run_forever(&mut self) {
+        loop {
+            let now = SystemTime::now();
+            match self.determine_action(now) {
+                Ok((batch_id, Action::AlreadyHandled)) => {
```

Comment:
If we're always waiting for batches to end, should this case really ever happen? If it shouldn't I would suggest it be an `Err` and not an `Ok`.

Reply:
I thought I'd leave this in in case the system time is changed while the scheduler is running. If it is changed backwards while we are waiting on the next batch then it can happen.